k2/csrc/fsa_algo.cc

```diff
@@ -274,6 +274,85 @@ bool Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map /*=nullptr*/) {
   return is_acyclic;
 }
 
+void RmEpsilonsPrunedMax(const WfsaWithFbWeights &a, float beam, Fsa *b,
+                         std::vector<float> *arc_weights_b) {
+  CHECK_GT(beam, 0);
+  CHECK_NOTNULL(b);
+  CHECK_NOTNULL(arc_weights_b);
+  b->arc_indexes.clear();
+  b->arcs.clear();
+  arc_weights_b->clear();
+
+  const auto &fsa = a.fsa;
+  if (IsEmpty(fsa)) return;
+  int32_t num_states = fsa.NumStates();
+  int32_t final_state = fsa.FinalState();
+  const auto &arcs_a = fsa.arcs;
+  const float *arc_weights_a = a.arc_weights;
+
+  // identify all states that should be kept
+  std::vector<char> non_eps_in(num_states, 0);
+  non_eps_in[0] = 1;
+  // (label, dest_state) -> `sum` of weights along all paths from current state
+  // to `dest_state` with label == `label`(or plus numbers of epsilon)
+  using ArcMap =
+      std::unordered_map<std::pair<int32_t, int32_t>, double, PairHash>;
+  std::vector<ArcMap> arcs_b(num_states);
+  for (int32_t i = static_cast<int32_t>(arcs_a.size()) - 1; i >= 0; --i) {
+    const auto &arc = arcs_a[i];
+    const auto src_state = arc.src_state;
+    const auto dest_state = arc.dest_state;
+    const auto label = arc.label;
+
+    double arc_weight = arc_weights_a[i];
+    if (label != kEpsilon) {
+      non_eps_in[dest_state] = 1;
+      auto insert_result = arcs_b[src_state].emplace(
+          std::make_pair(label, dest_state), arc_weight);
+      if (!insert_result.second) {
+        double &weight = insert_result.first->second;
+        // `sum` the weights of two paths
+        weight = std::max(weight, arc_weight);
+      }
+    } else {
+      // remove epsilon arcs
+      for (const auto &item : arcs_b[dest_state]) {
+        // `times` the arc weights along path
+        double new_weight = arc_weight + item.second;
+        auto insert_result = arcs_b[src_state].emplace(item.first, new_weight);
+        if (!insert_result.second) {
+          double &weight = insert_result.first->second;
+          // `sum` the weights of two paths
+          weight = std::max(weight, new_weight);
+        }
+      }
+    }
+  }
+  const double *forward_state_weights = a.ForwardStateWeights();
+  const double *backward_state_weights = a.BackwardStateWeights();
+  const double best_weight = forward_state_weights[final_state] - beam;
+  b->arc_indexes = fsa.arc_indexes;
+  auto &arc_indexes_b = b->arc_indexes;
+  int32_t arc_num_b = 0;
+  for (int32_t s = 0; s < b->NumStates(); ++s) {
+    arc_indexes_b[s] = arc_num_b;
+    if (non_eps_in[s] == 1) {
+      for (const auto &arc_map : arcs_b[s]) {
+        int32_t dest_state = arc_map.first.second;
+        double weight = arc_map.second;
+        if (forward_state_weights[s] + weight +
+                backward_state_weights[dest_state] >
+            best_weight) {
+          b->arcs.emplace_back(Arc{s, dest_state, arc_map.first.first});
```

Comment:
```cpp
b->arcs.emplace_back(s, dest_state, arc_map.first.first);
```

No need to write `Arc{` and `}`.  You have to rebase your change to avoid compiler errors.